### PR TITLE
RFD - Automatic Azure Server Discovery

### PR DESCRIPTION
**Pull request by atburke**
RFD for the automatic discovery and joining of Azure servers. This is essentially an Azure version of [RFD 57](https://github.com/gravitational/teleport/blob/master/rfd/0057-automatic-aws-server-discovery.md).

### CODE REVIEW

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (strideynet) [inline rfd/xxxx-automatic-azure-server-discovery.md:5]: Who are the required reviewers for this RFD?

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (strideynet) [inline rfd/xxxx-automatic-azure-server-discovery.md:57]: Should we identify which discovery service was responsible for discovering it ?

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (strideynet) [inline rfd/xxxx-automatic-azure-server-discovery.md:73]: Does this discovery only work if the discovery service is running within the same management group as the resources to be discovered?

**Review by strideynet — CHANGES_REQUESTED** [review]

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (tigrato) [inline rfd/xxxx-automatic-azure-server-discovery.md:10]: From what I understood what will discover is the Discovery service and not the SSH.

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (tigrato) [inline rfd/xxxx-automatic-azure-server-discovery.md:15]: What happens once Discovery finds a new server? Ignores it or will the RFD include the joining as well before implementing it?

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (tigrato) [inline rfd/xxxx-automatic-azure-server-discovery.md:38]: Discovery service

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (tigrato) [inline rfd/xxxx-automatic-azure-server-discovery.md:41]: How do you plan to add them? 
Currently, EC2 works with SSM, and nodes are only added to the cluster once they register themselves (after discovery service calls SSM to install Teleport on them).

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (tigrato) [inline rfd/xxxx-automatic-azure-server-discovery.md:54]: Can we leave this up to the user? 
If they want custom names we can check if they have a tag `teleport.dev/instance_name: {name}` like it's being done with AKS and Database discovery.

Adding a UUID to the name is quite difficult to follow and work with.

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (tigrato) [reply rfd/xxxx-automatic-azure-server-discovery.md:57]: Missing origin label

Comment (tigrato) [inline rfd/xxxx-automatic-azure-server-discovery.md:57]: Will you import VM labels into Teleport as well?
If yes, append the region and subID as well

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (zmb3) [inline rfd/xxxx-automatic-azure-server-discovery.md:22]:
```suggestion
RFD 57 replaced the process of manually installing Teleport on EC2 servers, which
could be slow for large numbers of instances. This RFD adds discovery of Azure instances.
The installation of Teleport on these servers is left to a future RFD.
```

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (zmb3) [inline rfd/xxxx-automatic-azure-server-discovery.md:31]: Is it legal to specify both `azure` and `aws` in the config for a single discovery service, or must you pick only one?

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (zmb3) [inline rfd/xxxx-automatic-azure-server-discovery.md:42]: I'm curious how we settled on 60 seconds here. Databases and desktops use a 5m discovery interval, for example.

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (tigrato) [reply rfd/xxxx-automatic-azure-server-discovery.md:31]: The AKS and EKS discovery already uses both at the same time but the service must have authentication credentials for both.

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (atburke) [reply rfd/xxxx-automatic-azure-server-discovery.md:114]: No. As long as it can see the resources to be discovered, the discovery service can be running anywhere.

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (atburke) [reply rfd/xxxx-automatic-azure-server-discovery.md:48]: I imagine it will be mostly the same Azure-equivalent process, but I think those details would fit better in the eventual registration RFD.

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (atburke) [reply rfd/xxxx-automatic-azure-server-discovery.md]: Yes, Teleport already imports Azure labels automatically.

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (jakule) [inline rfd/xxxx-automatic-azure-server-discovery.md:129]: Is that the only permission that we need for discovery?

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (jakule) [inline rfd/xxxx-automatic-azure-server-discovery.md:78]: ff? Fast Forward?

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (jakule) [reply rfd/xxxx-automatic-azure-server-discovery.md:49]: @zmb3 As every hardcoded timeout whatever value we pick there will be always questions why x not y. Is there any reason to not include the timeout in the config so it can be changed and use 5 min as a default?

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (atburke) [reply rfd/xxxx-automatic-azure-server-discovery.md:129]: It should be. All the discovery service needs to be able to do is get a list of virtual machines.

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (zmb3) [reply rfd/xxxx-automatic-azure-server-discovery.md:49]: This is not a timeout, but an interval for periodic operations. We mostly just use hard-coded constants for those. 

I don't feel strongly either way, 1m just feels pretty frequent for something like this.

**Review by jakule — APPROVED** [review]

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (tigrato) [inline rfd/xxxx-automatic-azure-server-discovery.md:123]: Include all fields defined in [AzureMatcher ](https://github.com/gravitational/teleport/blob/09773205f7fd96541e5d98ef0d2a790a8317a84e/lib/config/fileconf.go#L1357)
This is just to avoid confusion if anyone reads the RFD and doesn't see all the fields supported in the example.

**Review by tigrato — APPROVED** [review]

**Review by strideynet — APPROVED** [review]

**Review by r0mant — CHANGES_REQUESTED** [review]
@atburke @jakule This RFD should also cover how discovered instances will be enrolled in the Teleport cluster. On AWS we use SSM to run commands on the nodes, to my knowledge Azure does not provide anything similar. Will we need to use SSH, or is there any other mechanism by which you can run basically arbitrary commands on Azure VMs?

**Comment by tigrato** [comment]
azure has the `az vm run-command` family of commands.
They are similar to AWS SSM but you need to check if the machine is a Windows or Linux machine.
They are available via SDK as well

**Comment by atburke** [comment]
Reviewers, I've updated this RFD to include installation and would appreciate another round of reviews when you have time.

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (atburke) [reply rfd/xxxx-automatic-azure-server-discovery.md:117]: I haven't found an SSM equivalent in Azure. We'll have to send the commands in the `RunCommand` payload, same as AKS. I also haven't found a way to limit what commands can be run; the run command permissions seem to be all-or-nothing.

Comment (tigrato) [reply rfd/xxxx-automatic-azure-server-discovery.md:117]: From the docs [here](https://pkg.go.dev/github.com/azure/azure-sdk-for-go/services/compute/mgmt/2017-12-01/compute#RunCommandInput) I would say it should be possible - the command ID allows you to reference existing commands. The commands a user can invoke can be limited with `?scope?` field (I guess).

How to create and manage them is unknown to me

Comment (atburke) [reply rfd/xxxx-automatic-azure-server-discovery.md:117]: [Command ID is actually for specifying Azure built-in commands, not for referencing existing commands](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/run-command#available-commands). For any custom script, its value will always be `RunShellScript`.

Comment (tigrato) [reply rfd/xxxx-automatic-azure-server-discovery.md:117]: Gosh, it's even worst than I imagined.
Let's write that down and mention that it is impossible to control at the IAM level what can be executed by the Teleport identity. Add it to the security section

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (strideynet) [inline rfd/xxxx-automatic-azure-server-discovery.md:152]: Should the discovery labels be included here e.g `teleport.dev/discovered-by` or are these injected elsewhere ?

**Comment by strideynet** [comment]
Should we expect a separate RFD detailing joining on Azure ? Do we have an estimated timeline for that ?

**Comment by atburke** [comment]
No, joining is now a part of this RFD. I've updated to description to reflect that.

**Comment by zmb3** [comment]
@atburke @r0mant this RFD still has changes requested, but the implementation was already merged in #17850. Let's make sure the concerns here are addressed and try to get the RFD approved before we merge the code next time.

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (jakule) [inline rfd/xxxx-automatic-azure-server-discovery.md:117]: nit:
```suggestion
Azure auto-discovery will use the [Azure join method](https://github.com/gravitational/teleport/pull/19172).
```
Can you also link the RFD here?

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (jakule) [inline rfd/xxxx-automatic-azure-server-discovery.md:119]: I think we've already decided which we are using.

**Review by jakule — APPROVED** [review]
LGTM.

@r0mant Can you please take a look?

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (r0mant) [inline rfd/xxxx-automatic-azure-server-discovery.md:85]: Is this going to use the same default installer script as AWS discovery?

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (r0mant) [inline rfd/xxxx-automatic-azure-server-discovery.md:141]: Nit: I would clarify here again that this is not the static/dynamic token value but the name of the Azure IAM join token.

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (r0mant) [inline rfd/xxxx-automatic-azure-server-discovery.md:147]: Use `teleport.internal/` here.

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (r0mant) [inline rfd/xxxx-automatic-azure-server-discovery.md:153]: Is this actually parameterized/templated somehow or are you using ${} syntax just to indicate that this will use sub/VM IDs?

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (r0mant) [inline rfd/xxxx-automatic-azure-server-discovery.md:164]: `teleport.internal/`

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (r0mant) [inline rfd/xxxx-automatic-azure-server-discovery.md:132]: Can you give an example command Teleport will execute on VMs using this API, incl. parameters?

**Review by r0mant — APPROVED** [review]
lgtm with a few comments

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (atburke) [reply rfd/xxxx-automatic-azure-server-discovery.md:85]: I made a separate one while testing, but there's no reason we couldn't have one default installer that handles both.

File: rfd/xxxx-automatic-azure-server-discovery.md
Comment (atburke) [reply rfd/xxxx-automatic-azure-server-discovery.md]: Just to indicate those values. I've changed the syntax to make that clearer.

**Pull request merged**
